### PR TITLE
Circuit breaker async

### PR DESCRIPTION
**Pull request by danielmarbach**
Circuitbreaker Failure is async and cleanup of some `Thread.Sleep`

@Particular/tf-asyncawait Please review and merge

### CODE REVIEW

File: src/NServiceBus.Core/CircuitBreakers/RepeatedFailuresOverTimeCircuitBreaker.cs [hunk 1]
```diff
@@ -28,7 +29,7 @@ public void Success()
             Logger.InfoFormat("The circuit breaker for {0} is now disarmed", name);
         }
 
-        public void Failure(Exception exception)
+        public Task Failure(Exception exception)
```
Comment (SeanFeldman) [inline src/NServiceBus.Core/CircuitBreakers/RepeatedFailuresOverTimeCircuitBreaker.cs]: Should it be suffixed with Async?

File: src/NServiceBus.Core/DelayedDelivery/TimeoutManager/ExpiredTimeoutsPoller.cs [hunk 2]
```diff
@@ -59,17 +59,17 @@ public void Dispose()
         {
```
Comment (SeanFeldman) [inline src/NServiceBus.Core/DelayedDelivery/TimeoutManager/ExpiredTimeoutsPoller.cs]: Should it be suffixed with Async?

File: src/NServiceBus.AcceptanceTests/Basic/When_multiple_mappings_exists.cs [hunk 3]
```diff
@@ -55,8 +54,7 @@ public class MyMessageHandler : IHandleMessages<MyBaseCommand>
                 public Task Handle(MyBaseCommand message)
```
Comment (SeanFeldman) [inline src/NServiceBus.AcceptanceTests/Basic/When_multiple_mappings_exists.cs]: Not related to this PR - should `IHandleMessages<T>.Handle(T)` be suffixed with Async?

File: src/NServiceBus.Core/CircuitBreakers/RepeatedFailuresOverTimeCircuitBreaker.cs (hunk 1 above)
Comment (danielmarbach) [inline src/NServiceBus.Core/CircuitBreakers/RepeatedFailuresOverTimeCircuitBreaker.cs]: It is internal. So no



-- 
*#NSBcon 2015 • The "All About NServiceBus" conference is back *
*• Learn more <http://go.particular.net/l/27442/2015-08-21/5d4wzp> →*

File: src/NServiceBus.Core/DelayedDelivery/TimeoutManager/ExpiredTimeoutsPoller.cs (hunk 2 above)
Comment (danielmarbach) [inline src/NServiceBus.Core/DelayedDelivery/TimeoutManager/ExpiredTimeoutsPoller.cs]: It is internal. So no

File: src/NServiceBus.AcceptanceTests/Basic/When_multiple_mappings_exists.cs (hunk 3 above)
Comment (danielmarbach) [inline src/NServiceBus.AcceptanceTests/Basic/When_multiple_mappings_exists.cs]: Yes we can align that

File: src/NServiceBus.Core/CircuitBreakers/RepeatedFailuresOverTimeCircuitBreaker.cs (hunk 1 above)
Comment (SeanFeldman) [inline src/NServiceBus.Core/CircuitBreakers/RepeatedFailuresOverTimeCircuitBreaker.cs]: [MSDN disagrees](https://msdn.microsoft.com/en-us/library/hh191443.aspx?f=255&MSPPError=-2147217396#BKMK_NamingConvention) with you :grinning: 

In the [Example](https://msdn.microsoft.com/en-us/library/hh191443.aspx?f=255&MSPPError=-2147217396#BKMK_CompleteExample) `AccessTheWebAsync()` is internal async method and still prefixed. I was under impression that access level doesn't matter, intent matters.

Comment (SimonCropp) [inline src/NServiceBus.Core/CircuitBreakers/RepeatedFailuresOverTimeCircuitBreaker.cs]: @SeanFeldman yeah MS released that advice under the misconception that people would be implementing both sync and async API in the same lib. turns out "both sync and async" means you need to basically need to implement the entire stack twice. as such the general advice now is to "if it is IO bout only do async and do not suffix"

Comment (danielmarbach) [inline src/NServiceBus.Core/CircuitBreakers/RepeatedFailuresOverTimeCircuitBreaker.cs]: @SeanFeldman I think we talked enough about that, you were also part of that discussion. Our decision was: On the move towards the end goal we postfix code which is directly used by customers with `Async`. In the end goal we will remove those postfixes

**Comment by danielmarbach** [comment]
Rebased. I'm going to merge since reviews by @SeanFeldman didn't find anything blocking a merge

**Pull request merged**